File: haskell/Powercards/GamePlay.hs
module GamePlay where

import Core
import qualified Utils as U
import qualified Select as S
import Control.Monad (foldM_)
import Data.Maybe (fromJust, isJust)
import Data.Graph.Inductive.Query.Monad (mapSnd)

createGame :: [String] -> Game
createGame [] = error "must have at least 1 player"
createGame playerNames = addGlobals . addPlayers $ emptyGame
    where addPlayers :: Game -> Game
          addPlayers game = foldr addPlayer game $ zip [0..] playerNames

          addPlayer :: (Int, String) -> Game -> Game
          addPlayer (n, playerName) = 
            set (pDeck n) (drop 5 startDeck) .
            set (pHand n) (take 5 startDeck ++ [remodelCard, throneRoomCard]) .
            set (pPlayed n) [] .
            set (pDiscard n) [] .
            set (pName n) playerName

          addGlobals :: Game -> Game
          addGlobals = 
            set gTrash [] .
            set gStage Action .
            set gActivePlayerIndex 0 .
            set gPlayerCount (length playerNames) .
            set gActions 1 .
            set gBuys 1 .
            set gCoins 0

          startDeck :: [Card]
          startDeck = replicate 3 estateCard ++ replicate 7 copperCard

play :: Game -> IO ()
play game = foldM_ playOne game [1..]

playOne :: Game -> Int -> IO Game
playOne game _ | get gStage game == Action && get gActions game == 0 =
    return . set gStage Treasure $ game
playOne game _ | get gStage game == Action && get gActions game > 0 =
    S.select 
        (get aHand game) 
        (isJust . toActionCard)
        S.OptionalOne 
        (get aName game ++ ": select an action card to play")
    >>= handler
    where handler [] = do
            putStrLn "skip to treasure stage"
            return . set gStage Treasure . set gActions 0 $ game
          handler indexes@[_] =
            playCard . mapSnd (update gActions pred) . moveCards aHand aPlayed indexes $ game
          playCard ([card], game) = 
            putStrLn ("playing " ++ show card) >>
            (fromJust . toActionCard) card game
playOne game _ | get gStage game == Treasure =
    S.select 
        (get aHand game) 
        (isJust . toTreasureCard)
        S.Unlimited
        (get aName game ++ ": select treasure cards to play")
    >>= handler
    where handler [] = do
            putStrLn "skip to buy stage"
            return . set gStage Buy $ game
          handler indexes =
            gainCoin . moveCards aHand aPlayed indexes $ game
          gainCoin (cards, game) = do
            let coins = sum $ map (fromJust . toTreasureCard) cards
                game' = update gCoins (+coins) game
            putStrLn ("gained " ++ show coins ++ " coins")
            return game'
playOne game _ | get gStage game == Buy && get gBuys game == 0 =
    return . set gStage Cleanup $ game
playOne game _ | get gStage game == Buy && get gBuys game > 0 = do
    putStrLn "skipping buy stage as not implemented"
    return . set gStage Cleanup $ game
playOne game _ | get gStage game == Cleanup =
    return . reset . cleanup $ game
    where cleanup = drawCards aDeck aHand aDiscard 5 .
                    snd . moveCards aDiscard aPlayed [0..] . 
                    snd . moveCards aHand aPlayed [0..]
          reset = update gActivePlayerIndex nextApi . 
                  set gActions 1 . set gBuys 1 . set gCoins 0 . 
                  set gStage Action
          nextApi api = cycle [0..(get gPlayerCount game - 1)] !! (api + 1)

moveCards :: BoxAccess [Card] -> BoxAccess [Card] -> [Int] -> Game -> ([Card], Game)
moveCards ba1 ba2 indexes game =
    let v1 = get ba1 game
        (selected, unselected) = U.divide v1 indexes
        withdraw = set ba1 unselected
        deposit = update ba2 (++selected)
        pack g = (selected, g)
    in  pack . deposit . withdraw $ game

drawCards:: BoxAccess [Card] -> BoxAccess [Card] -> BoxAccess [Card] -> Int -> Game -> Game
drawCards _ _ _ 0 game = game
drawCards src dst bak n game =
    let (cards', game') = moveCards src dst [0..(n-1)] game
    in  case n - length cards' of 
        0 -> game'
        n' | n' > 0 -> snd . moveCards src dst [0..(n'-1)] . 
                       snd . moveCards bak src [0..] $ game'

trashCard :: Game -> IO (Maybe Card, Game)
trashCard game =
    S.select
        (get aHand game) 
        (const True)
        S.MandatoryOne
        "select a card to trash"
    >>= handler
    where handler [] = 
            putStrLn "No card to trash" >>
            return (Nothing, game)
          handler indexes@[_] =
            let ([card], game') = moveCards aHand gTrash indexes game
            in  putStrLn ("trashed " ++ show card) >> return (Just card, game')

gainCard :: Game -> IO (Maybe Card, Game)
gainCard game =
    S.select 
        supply
        (const True)
        S.MandatoryOne
        "select a card to gain"
    >>= handler
    where handler [] =
            putStrLn "No card to gain" >>
            return (Nothing, game)
          handler indexes@[_] =
            let (selected@[card], _) = U.divide supply indexes
                game' = update aDiscard (++selected) game
            in  putStrLn ("gained " ++ show card) >> return (Just card, game') 
          supply = [remodelCard, throneRoomCard]

estateCard :: Card
estateCard = BasicVictoryCard "Estate" 2 1

duchyCard :: Card
duchyCard = BasicVictoryCard "Duchy" 5 3

provinceCard :: Card
provinceCard = BasicVictoryCard "Province" 8 6

copperCard :: Card
copperCard = BasicTreasureCard "Copper" 0 1

silverCard :: Card
silverCard = BasicTreasureCard "Silver" 3 2

goldCard :: Card
goldCard = BasicTreasureCard "Gold" 6 3

remodelCard :: Card
remodelCard = BasicActionCard "Remodel" 4 cardPlay
    where cardPlay = (>>= handler) . trashCard
          handler (Just _, game) = fmap snd $ gainCard game
          handler (Nothing, game) = return game

throneRoomCard :: Card
throneRoomCard = BasicActionCard "Throne Room" 4 cardPlay
    where cardPlay game =
            S.select 
                (get aHand game) 
                (isJust . toActionCard)
                S.MandatoryOne 
                "select an action card to play twice"
            >>= handler
            where handler [] =
                    putStrLn "No action card to play" >>
                    return game
                  handler indexes@[_] =
                    playCardTwice . moveCards aHand aPlayed indexes $ game
                  playCardTwice ([card], game) = do
                    putStrLn ("playing " ++ show card ++ " first time")
                    game' <- (fromJust . toActionCard) card game
                    putStrLn ("playing " ++ show card ++ " second time")
                    (fromJust . toActionCard) card game'

File: python/powercards/gameplay.py
import dialog
import random

class Game(object):
    def __init__(self, player_names):
        assert 2 <= len(player_names) <= 4, 'only 2 to 4 players allowed'
        
        self.players = [Player(name) for name in player_names]
        self.active_index = random.randint(0, len(self.players)-1)
        self.board = Board([(Copper, 60), (Estate, 12)])
        self.stage = 'action'

        self.active_player().actions = 1

    def active_player(self):
        return self.players[self.active_index]

class Player(object):
    def __init__(self, name):
        self.name = name
        deck = [Estate() for i in range(3)] + [Copper() for i in range(7)]
        random.shuffle(deck)
        self.deck, self.hand = deck[:5], deck[5:]
        self.played, self.discard = [], []
        self.actions, self.buys, self.coins = 0, 0, 0

class Board(object):
    def __init__(self, piles):
        self.trash = []
        self.piles = piles

class Pile(object):
    def __init__(self, factory, count):
        self.factory = factory
        self.remaining = count
        self.sample = self.factory()

    def empty(self):
        return self.remaining <= 0

    def push(self):
        self.remaining += 1

    def pop(self):
        assert not self.empty, '{} pile is empty'.format(self.sample.name)
        self.remaining -= 1
        return self.factory()

    def name(self):
        return self.sample.name

def move_one(src, index, dst):
    card = src[index]
    del src[index]
    dst.append(card)
    return card

def move_many(src, indexes, dst):
    cards = [src[i] for i in indexes]
    cards_rev = [card for i, card in enumerate(src) if i not in indexes]
    src.clear()
    src.extend(cards_rev)
    dst.extend(cards)
    return cards

def discard_deck(player):
    player.discard.extend(player.deck[::-1])
    player.deck = []

def putback_many_to_deck(src, indexes, dst):
    return move_many(src, indexes, dst)

def trash_one(src, index, game):
    return move_one(src, index, game.board.trash)

def gain_one(pile, dst):
    card = pile.pop
    dst.append(card)
    return card

def play_action(game):
    if game.active_player().actions == 0:
        dialog.out('no more actions, skip to treasure stage')
        game.stage = 'treasure'
    else:
        chosen = dialog.optional_one(
            message='select an action card to play',
            items=[(card.name, isinstance(card, ActionCard)) for card in game.active_player().hand])
        if chosen is None:
            dialog.out('skip to treasure stage')
            game.stage = 'treasure'
        else:
            dialog.out('you have chosen {0}'.format(chosen))

def play_treasure(game):
    chosen = dialog.unlimited(
        message='select treasure cards to play',
        items=[(card.name, isinstance(card, TreasureCard)) for card in game.active_player().hand])
    if chosen is None:
        dialog.out('skip to buy stage')
        game.stage = 'buy'
    else:
        dialog.out('you have chosen {0}'.format(chosen))

def play_buy(game):
    pass

def play_cleanup(game):
    pass

def play(game):
    if game.stage == 'action':
        play_action(game)
    elif game.stage == 'treasure':
        play_treasure(game)
    elif game.stage == 'buy':
        play_buy(game)
    elif game.stage == 'cleanup':
        play_cleanup(game)

class Card(object):
    def name(self):
        return type(self).__name__

class ActionCard(Card):
    pass

class TreasureCard(Card):
    pass

class VictoryCard(Card):
    pass

class Copper(TreasureCard):
    pass

class Estate(VictoryCard):
    pass

class Remodel(ActionCard):
    costs = 4

    def play(game):
        trash_i = dialog.one(
            message='select a card to remodel',
            items=[(card.name, True) for card in game.active.hand])
        if trash_i is None:
            dialog.out('no card in hand to remodel')
        else:
            trashed_card = trash_one(game.active_player().hand, trasn_i, game)
            dialog.out('trashed {}'.format(trashed_card.name))

            gain_i = dialog.one(
                message='select a pile to gain',
                items=[(pile.name, not pile.empty and pile.sample.costs <= trashed_card.costs + 2) for pile in game.board.piles])
            if gain_i is None:
                dialog.out('no pile available to gain')
            else:
                gained_card = gain_one(game.board.piles[gain_i], game.active_player().discard)
                dialog.out('gained {}'.format(gained_card.name))

